Fix map linting crashing for not existing directories

abcdefg30 opened the pull request:
Fixes the regression that slipped in #13938 in (see https://github.com/OpenRA/OpenRA/pull/13938#issuecomment-334919096).

For reference: https://logs.openra.net/?year=2017&month=08&day=30#15:42:38

abcdefg30 committed:
Fix map linting crashing for not existing directories

abcdefg30 committed:
Use MapCache.LoadMapsWithoutCaching in UpgradeModCommand

Phrohdoh commented on OpenRA.Game/Map/MapCache.cs:
```diff
@@ -122,6 +122,47 @@ public void LoadMaps()
+			{
+				var name = kv.Key;
+				var classification = string.IsNullOrEmpty(kv.Value)
+					? MapClassification.Unknown : Enum<MapClassification>.Parse(kv.Value);
```
`Enum.TryParse` and list the valid values upon failure.

Phrohdoh replied:
Ignore this I suppose. You're just moving code afterall.

pchote replied:
This would be a good opportunity to fix the original code (since this is pulling out a common helper).

Phrohdoh commented on OpenRA.Mods.Common/UtilityCommands/UpgradeModCommand.cs:
```diff
@@ -78,42 +77,19 @@ void IUtilityCommand.Run(Utility utility, string[] args)
 			// Only upgrade system maps - user maps must be updated manually using --upgrade-map
 			Console.WriteLine("Processing Maps:");
-			foreach (var kv in modData.Manifest.MapFolders)
+			var maps = modData.MapCache.EnumerateMapsWithoutCaching();
```
Why store this in a local?

abcdefg30 commented:
Updated.

pchote commented on OpenRA.Game/Map/MapCache.cs:
```diff
@@ -122,6 +122,48 @@ public void LoadMaps()
 			}
 		}
 
+		public IEnumerable<Map> EnumerateMapsWithoutCaching(MapClassification classification = MapClassification.System)
```
For this to work properly you will need to change the `MapClassification` definition to:
```csharp
// Used for grouping maps in the UI and for filtering
[Flags]
public enum MapClassification
{
	Unknown = 0,
	System = 1,
	User = 2,
	Remote = 4
}
```

pchote commented on OpenRA.Game/Map/MapCache.cs:
```diff
@@ -122,6 +122,48 @@ public void LoadMaps()
+				if (!Enum.TryParse(kv.Value, out packageClassification))
+					continue;
+
+				if (packageClassification != classification)
```
This will then need to become
```csharp
if (!classification.HasFlag(packageClassification))
```

pchote commented on OpenRA.Mods.Common/UtilityCommands/CheckYaml.cs:
```diff
@@ -77,29 +77,8 @@ void IUtilityCommand.Run(Utility utility, string[] args)
-						}
-					}
+					// Use all system maps for lint checking
+					maps = modData.MapCache.EnumerateMapsWithoutCaching().ToList();
```
This is silently changing the command behaviour.

Please do `EnumerateMapsWithoutCaching(MapClassification.System | MapClassification.User)` to keep the current behaviour for prep.

pchote replied:
This can be changed to enumerate only system dirs in a followup commit that is not cherry-picked.

pchote commented on OpenRA.Mods.Common/UtilityCommands/UpgradeModCommand.cs:
```diff
@@ -78,42 +77,18 @@ void IUtilityCommand.Run(Utility utility, string[] args)
-							}
-						}
-					}
+					Console.WriteLine(map.Package.Name);
```
Left over debugging code?

abcdefg30 replied:
This is intentional. We also print out which yaml files are currently upgraded. It also takes quite a while to update all maps, and you're left without knowing if something even happens.

pchote commented on OpenRA.Game/Map/MapCache.cs:
```diff
@@ -122,6 +122,48 @@ public void LoadMaps()
+				{
+					// Assume that the path is a directory if there is not an existing file with the same name
+					var resolved = Platform.ResolvePath(name);
+					if (!File.Exists(resolved))
```
The support directory is not a file, so this always fails.  This needs to also check `Directory.Exists`.  The comment also needs to be corrected.

abcdefg30 committed:
Log which maps are processed by UpgradeModCommand

Upgrading the maps can take quite some time,
and at the moment you are left without information if something is happening.

abcdefg30 commented:
Updated.

abcdefg30 committed:
Only lint system maps

pchote commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/a07639020e4439b0170edb759e9c50c8dc7754f2
